src/include/kompute/Image.hpp

```diff
@@ -0,0 +1,469 @@
+// SPDX-License-Identifier: Apache-2.0
+#pragma once
+
+#include "Memory.hpp"
+#include "Tensor.hpp"
+#include "kompute/Core.hpp"
+#include "logger/Logger.hpp"
+#include <memory>
+#include <string>
+
+namespace kp {
+
+/**
+ * Image data used in GPU operations.
+ *
+ * Each image would has a respective Vulkan memory and image, which
+ * would be used to store their respective data. The images can be used for GPU
+ * data storage or transfer.
+ */
+class Image : public Memory
+{
+  public:
+    enum class ImageDataTypes
+    {
+        eS8 = 0,
+        eU8 = 1,
+        eS16 = 2,
+        eU16 = 3,
+        eS32 = 4,
+        eU32 = 5,
+        eF16 = 6,
+        eF32 = 7,
+    };
+
+    static std::string toString(ImageDataTypes dt);
+
+    static Tensor::TensorDataTypes getTensorDataType(ImageDataTypes dt);
+
+    /**
+     *  Constructor with data provided which would be used to create the
+     *  respective vulkan image and memory.
+     *
+     *  @param physicalDevice The physical device to use to fetch properties
+     *  @param device The device to use to create the image and memory from
+     *  @param data Non-zero-sized vector of data that will be used by the
+     * image
+     *  @param width Width of the image in pixels
+     *  @param height Height of the image in pixels
+     *  @param numChannels The number of channels in the image
+     *  @param dataType Data type for the image which is of type ImageDataTypes
+     *  @param memoryType Type for the image which is of type MemoryTypes
+     *  @param tiling Tiling mode to use for the image.
+     */
+    Image(std::shared_ptr<vk::PhysicalDevice> physicalDevice,
+          std::shared_ptr<vk::Device> device,
+          void* data,
+          uint32_t width,
+          uint32_t height,
+          uint32_t numChannels,
+          const ImageDataTypes& dataType,
+          vk::ImageTiling tiling,
```

Comment:
I didn't do this because there is a version below which doesn't take a tiling parameter. It infers what the tiling should be based on the memory type (host-accessible memory will be linear tiled, device-only memory will be optimal tiled).

Reply:
Understood